### PR TITLE
[MRG+1] Respect custom log level: fixes GH-1612

### PR DESCRIPTION
**Pull request by lopuhin**
See #1612: ~~this uses ``self.settings`` instead of ``settings``, as elsewhere in the crawler init.~~
Main change is that now a new root logger is installed when a crawler is created if one was already installed before. This allows to respect custom settings related to logging, such as LOG_LEVEL, LOG_FILE, etc.

As the result, in case of several spiders, logging settings from the last spider take precedence. Without this change, all logging settings in ``custom_settings`` are ignored. Both behaviours are not great, but it seems that in most cases there is just one spider running in the process, so the new behaviour seems less bad, but it's a tough call.

### CODE REVIEW

**Comment by kmike** [comment]
The fix makes sense, but I wonder what happens if there are two spiders in the same process (executed via CrawlerRunner), and one of them changes LOG_LEVEL - does it affect the other spider?

**Comment by lopuhin** [comment]
Yes, I don't think you can have two spiders with different log levels in one process, and I don't immediately see a backwards-compatible fix for that.

**Comment by djunzu** [comment]
I guess the last spider to start will override any previous log level.
In this case I think a warning note should be somewhere in the docs saying it.

**Comment by kmike** [comment]
@djunzu I'd expect it to still follow log level for this spider, at least if ``spider`` is in context. This is not always possible though because some messages are global (without ``spider`` in context). But if we document current behavior then maybe we'll documet a bug; I'm not sure. Anyways, it looks like a different issue.

**Comment by djunzu** [comment]
@kmike I decided to test it just to know what happens. I tried
```python
# settings.py:
LOG_LEVEL = 'WARNING'

#testspider.py:
class TestSpider(scrapy.Spider):
    name = "Test"
    custom_settings = {
        'LOG_LEVEL': 'DEBUG',
    }

# output just with warning messages
```
and
```python
# settings.py:
LOG_LEVEL = 'DEBUG'

#testspider.py:
class TestSpider(scrapy.Spider):
    name = "Test"
    custom_settings = {
        'LOG_LEVEL': 'WARNING',
    }

# output with a lot of debug messages
```
and
```python
# settings.py:
#LOG_LEVEL = 'DEBUG'

#testspider.py:
class TestSpider(scrapy.Spider):
    name = "Test"
    custom_settings = {
        'LOG_LEVEL': 'WARNING',
    }

# output with a lot of debug messages (default LOG_LEVEL)
```
None of them works. It does not solve the problem and it is still not possible to set log level using `custom_settings` attribute. (This or I did something very wrong!)

tested with scrapy 1.3.2

@lopuhin Did you do any other test than the one you wrote?

**Comment by lopuhin** [comment]
@djunzu yes, I did check that this pull request solves the issue for me (specifically, I was using ``scrapy runspider`` command), and I checked that the test I wrote fails without this PR. Let me also check the ``scrapy crawl`` command.

**Comment by lopuhin** [comment]
@djunzu you are right, this is my mistake. Honestly, I'm not sure how did I test it and get what I expected. Thanks for checking! I'll see if I can fix it.
The test also needs updating: it just check log stats (which do respect log level in custom_settings), but not the actual log output.

**Comment by lopuhin** [comment]
The tests are green finally, yay! @kmike @djunzu do you mind having another look at this? I updated the PR description with summary of changes and also checked that ``scrapy crawl`` and ``scrapy runspider`` commands both respect custom log settings.

File: scrapy/crawler.py [hunk 1]
```diff
@@ -35,8 +35,9 @@ def __init__(self, spidercls, settings=None):
         self.signals = SignalManager(self)
         self.stats = load_object(self.settings['STATS_CLASS'])(self)
 
-        handler = LogCounterHandler(self, level=settings.get('LOG_LEVEL'))
+        handler = LogCounterHandler(self, level=self.settings.get('LOG_LEVEL'))
         logging.root.addHandler(handler)
+        configure_logging(self.settings)
```
Comment (kmike) [inline scrapy/crawler.py]: Hm, previously this function was only called by Scrapy commands, not by Crawler. I think we should check carefully that calling it in Crawler doesn't affect usage examples in https://doc.scrapy.org/en/latest/topics/logging.html#scrapy.utils.log.configure_logging.

Comment (lopuhin) [inline scrapy/crawler.py]: Right, thanks for the catch! Probably it's wrong to call it unconditionally when starting any crawler - when you are using scrapy as a library you'd like to be in control of the logging.  Initially I added this call to support other log settings as well, but now I see that this is problematic. Maybe it's better to change scrapy root handler only if it's already installed.

**Comment by lopuhin** [comment]
Now a new root logger is installed when a crawler is created only if one was already installed before.

**Comment by expansion** [comment]
[https://mailguru.s3.amazonaws.com/images/spark/qr.v2/935cf1c585724824a8e35714de5e0627.png]

谢谢！



Spark by Readdle<https://itunes.apple.com/app/id997102246?mt=8&uo=4&at=10l6UJ&ct=Spark_QR_viral>

        Spark by Readdle<https://itunes.apple.com/app/id997102246?mt=8&uo=4&at=10l6UJ&ct=Spark_QR_viral>

File: tests/test_crawl.py [hunk 2]
```diff
@@ -97,8 +97,8 @@ def test_retry_dns_error(self):
 
     @defer.inlineCallbacks
     def test_start_requests_bug_before_yield(self):
+        crawler = self.runner.create_crawler(BrokenStartRequestsSpider)
         with LogCapture('scrapy', level=logging.ERROR) as l:
-            crawler = self.runner.create_crawler(BrokenStartRequestsSpider)
```
Comment (kmike) [inline tests/test_crawl.py]: what is the reason for this change?

Comment (lopuhin) [inline tests/test_crawl.py]: I made that changes because a new root log handler was installed when crawler was created, and so LogCapture did not work (and many other tests also have crawler creation outside of LogCapture). But it looks like it is no longer required, at least when the tests are run in isolation. I revered this changes in 6f55ca4, let's see if the full test suite passes.

Comment (lopuhin) [inline tests/test_crawl.py]: Yep, the tests passed, probably it's better to leave it as it was in master. Thanks for the catch!

**Comment by kmike** [comment]
Looks good to me, thanks @lopuhin for the fix and @djunzu for the review.

**Comment by djunzu** [comment]
:+1: 


Back to my initial comment:
I think that docs should have this note: "in case of several spiders, logging settings from the last spider take precedence".

@kmike, regarding
> But if we document current behavior then maybe we'll documet a bug

I think documented bugs are good: everyone knows what to expect. Undocumented bugs are bad: when you are not expecting it, something blows up in front of you!

Nonetheless, +1 for merge.
